### PR TITLE
[FIX] web_editor: keep cursor in empty inline tag

### PR DESCRIPTION
**Pull request by sebgeelen**
Ensure the user can keep typing inside a newly emptied inline tag.

Some changes had to be made to the deleteForward and backward commands,
to ensure the new behaviors.

task-2622455

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 1]
```diff
@@ -85,6 +85,40 @@ describe('Editor', () => {
...
+                    await testEditor(BasicEditor, {
+                        contentBefore: '<p><span>[]\u200B</span><span>ab</span></p>',
+                        stepFunction: deleteForward,
+                        contentAfter: '<p><span>[]b</span></p>',
+                    });
+                    await testEditor(BasicEditor, {
+                        contentBefore: '<p><span>[]\u200B</span><br><b>cd</b></p>',
+                        stepFunction: deleteForward,
+                        contentAfter: '<p>[]<b>cd</b></p>',
+                    });
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:120]: Oh my, those tests are sexy !

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 2]
```diff
@@ -753,12 +787,62 @@ X[]
...
+                    },
+                    contentAfter: '<p>ab []ef</p>',
+                });
+                await testEditor(BasicEditor, {
+                    contentBefore: '<p>uv<span class="style-class">[wx]</span>yz</p>',
+                    stepFunction: async editor => {
+                        await deleteForward(editor);
+                        await deleteForward(editor);
+                    },
+                    contentAfter: '<p>uv[]z</p>',
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:845]: Damn those ones too ! :p

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 3]
```diff
@@ -963,6 +1047,43 @@ X[]
...
+                        contentBefore: '<p>ab\u200B[]c</p>',
+                        stepFunction: deleteBackward,
+                        contentAfter: '<p>a[]c</p>',
+                    });
+                });
+                it('should keep inline block', async () => {
+                    await testEditor(BasicEditor, {
+                        contentBefore: '<div><p>ab</p><br><span>c[]</span></div>',
+                        stepFunction: deleteBackward,
+                        contentAfter: '<div><p>ab</p><br>[]</div>',
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:1061]: Would it have kept the span if there was a style on it ? That seems important.

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 4]
```diff
@@ -963,6 +1047,43 @@ X[]
...
+                    });
+                    await testEditor(BasicEditor, {
+                        contentBefore: '<div><p>cd</p><br><span>e[]</span></div>',
+                        stepFunction: async editor => {
+                            await deleteBackward(editor);
+                            await insertText(editor, 'x');
+                        },
+                        contentAfter: '<div><p>cd</p><br><span>[]x</span></div>',
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:1069]: I guess this kinda answer my question... but how comes we don't see the span in the previous test ? Because of cleanForSave ?

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 5]
```diff
@@ -1965,7 +2086,7 @@ X[]
                 await testEditor(BasicEditor, {
                     contentBefore: '<p>ab<b class="oe_unremovable">[cd]</b>ef</p>',
                     stepFunction: deleteBackward,
-                    contentAfter: '<p>ab<b class="oe_unremovable">[\u200B]</b>ef</p>',
+                    contentAfter: '<p>ab<b class="oe_unremovable">[]\u200B</b>ef</p>',
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:2089]: This is so much better !

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js [hunk 6]
```diff
@@ -3091,12 +3212,12 @@ X[]
                 await testEditor(OdooEditor, {
                     contentBefore: '<p>[]</p>',
                     stepFunction: async editor => {
-                        insertText(editor, 'a');
-                        insertText(editor, 'b');
-                        insertText(editor, 'c');
+                        await insertText(editor, 'a');
+                        await insertText(editor, 'b');
+                        await insertText(editor, 'c');
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:3217]: How comes it used to work before ?

File: addons/web_editor/static/lib/odoo-editor/test/spec/utils.test.js [hunk 7]
```diff
@@ -35,7 +35,15 @@ import {
 } from '../../src/utils/utils.js';
 import { BasicEditor, testEditor } from '../utils.js';
 
+const cleanTestHtml = () => {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/spec/utils.test.js:38]: I guess this should also go in the "clean feature" commit ?

File: addons/web_editor/static/lib/odoo-editor/test/utils.js [hunk 8]
```diff
@@ -287,6 +287,7 @@ export function renderTextualSelection() {
 
 export async function testEditor(Editor = OdooEditor, spec) {
     const testNode = document.createElement('div');
+    document.querySelector('#editor-test-container').innerHTML = '';
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/utils.js:290]: Why is this needed now ?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 9]
```diff
@@ -1340,8 +1339,7 @@ export class OdooEditor extends EventTarget {
             const el = closestElement(joinWith);
             const { zws } = fillEmpty(el);
             if (zws) {
-                // ZWS selection in OdooEditor is not working in current version of firefox (since v93.0)
-                setSelection(zws, 0, zws, this.isFirefox ? 0 : nodeSize(zws));
+                setSelection(zws, 0, zws, 0);
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:1342]: I was just using the firefox fix everywhere.
But after rethinking it, we loose some advantages if we do not select the ZWS. 
It can generate glitch when keyboard navigating around ZWS, so selecting it can avoid the first of theses instances.

So I'll revert this

File: addons/web_editor/static/lib/odoo-editor/src/commands/deleteForward.js [hunk 10]
```diff
@@ -14,21 +13,78 @@ import {
...
+
+    if (offset === this.nodeValue.length) {
+        // Delete at the end of a text node is not a specific case to
+        // handle, let the element implementation handle it.
+        HTMLElement.prototype.oDeleteForward.call(this, offset, alreadyMoved);
+        return;
     }
+
+    // Get the size of the unicode character to remove.
+    const charSize = [...this.nodeValue.slice(0, offset + 1)].pop().length;
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/commands/deleteForward.js:37]: Honestly I don't know. I just "copied" this code sniped from `deleteBackward.js > ln:38`.

I haven't tested it. 
But it should work as well as before, since I've reproduced the logic that was previously applied in `Text.prototype.oDeleteBackward` into `Text.prototype.oDeleteForward`.

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js (hunk 4 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:1069]: yes, because clean for save.

We indeed keep an empty inline tag after the clean for save, if it have a class.
I'll add a more explicit test about that.

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js (hunk 5 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:2089]: well ... 
https://github.com/odoo/odoo/pull/84187#discussion_r803649249

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js (hunk 6 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:3217]: I'm not sure the awaits are necessary in fact.
I've added them during my debuging to check if it was not the source of one of my problem. 
I just didn't remove them...

File: addons/web_editor/static/lib/odoo-editor/test/utils.js (hunk 8 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/utils.js:290]: Sometimes some tests are affected if the previous Editor instances and it's editable DIV are not completely and properly cleaned. 
e.g. test **green** if run alone, but **red** in bulk.

No idea why, but I added that to make sure it didn't append.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js]: After Discord discussion, we decided to keep the range collapsed. mainly so the user can always see the cursor in the document.

File: addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js (hunk 3 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/spec/editor.test.js:1100]: yes

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js [hunk 11]
```diff
@@ -1,5 +1,5 @@
 /** @odoo-module **/
-const INVISIBLE_REGEX = /\u200c/g;
+const INVISIBLE_REGEX = /\u200B/g;
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2]: C'est bizarre quand même que c'était `c` avant non ? Cela dit elle n'est même plus utilisée cette constante ?

File: addons/web_editor/static/lib/odoo-editor/test/editor-test.js [hunk 12]
```diff
@@ -25,6 +25,35 @@ mocha.run(failures => {
...
+        const cleanUl = function () {
+            const emptyUl = report.querySelectorAll('ul:empty');
+            for (let ul of emptyUl) {
+                ul.parentElement.remove();
+            }
+        };
+        cleanUl();
+        cleanUl();
+        cleanUl();
+        cleanUl();
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/editor-test.js:57]: Because the UL are multiple and nested in the test report, so I call the method multiple time to clean all unnecessary title. It's crude but is not really a critical feature so I did not bother make it look cleaner.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 13]
```diff
@@ -1240,6 +1240,14 @@ export class OdooEditor extends EventTarget {
         if (!range) return;
         let start = range.startContainer;
         let end = range.endContainer;
+        const contentIsZWS = range.cloneContents().textContent === '\u200B';
+        if (start === end && end.nodeType === Node.TEXT_NODE && contentIsZWS) {
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:1244]: Should be possible, yes.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 14]
```diff
@@ -2155,12 +2174,25 @@ export class OdooEditor extends EventTarget {
...
+                // Unit tests events are not trusted by the browser,
+                // the insertText has to be done manualy.
+                const isUnitTests = !ev.isTrusted && this.testMode;
+                // we cannot trust the browser to keep the selection inside empty tags.
+                const latestSelectionInsideEmptyTag = this._isLatestComputedSelectionInsideEmptyInlineTag();
+                if (wasTextSelected || isUnitTests || latestSelectionInsideEmptyTag) {
                     ev.preventDefault();
-                    this._applyRawCommand('oDeleteBackward');
+                    if (!isUnitTests) {
+                        // We need to undo the character inserted by the browser first
```
[the remark this answers is not in the document]
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2186]: I'll explain why the unit tests do not need the deletebackward

File: addons/web_editor/static/lib/odoo-editor/src/utils/utils.js (hunk 11 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:2]: C'est bien possible quelle n'est plus utilisée en effet, je vais vérifier

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+

File: addons/web_editor/static/lib/odoo-editor/test/utils.js [hunk 15]
```diff
@@ -134,35 +134,6 @@ export function parseMultipleTextualSelection(testContainer) {
...
 }
 
-/**
- * Insert in the DOM:
- * - `SELECTION_ANCHOR_CHAR` in place for the selection start
- * - `SELECTION_FOCUS_CHAR` in place for the selection end
- *
- * This is used in the function `testEditor`.
- */
-export function renderMultipleTextualSelection() {
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/test/utils.js:144]: Elle est partie où cette fonction là ?

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js [hunk 16]
```diff
@@ -2302,8 +2360,53 @@ export class OdooEditor extends EventTarget {
...
+    /**
+     * initialise the provided element to be ready for edition
+     *
+     */
+    initElementForEdition(element = this.editable) {
+        // Flag elements with forced contenteditable=false.
+        // We need the flag to be able to leave the contentEditable
+        // at the end of the edition (see cleanForSave())
+        for (const el of element.querySelectorAll('[contenteditable="false"]')) {
+            el.setAttribute("oe-keep-contenteditable", '1');
```
Comment (dmo-odoo) [inline addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2373]: On peut pas avoir des `true` plutôt ? Pour ça et l'autre avec les zws là. Je trouve que ça fait un peu bizarre ces '1' magiques qui se baladent dans le code... ^^'

File: addons/web_editor/static/lib/odoo-editor/test/utils.js (hunk 15 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/test/utils.js:144]: Elle était jamais utilisée, et c'est la copie parfaite de `parseTextualSelection`. :shrug: 
Donc je l'ai virée.

File: addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js (hunk 16 above)
Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2373]: Ben c'est obligé d'être une string, vu que ça va dans un attributs HTML. 
Ça risque d'être ambigus de devoir faire `attribute === 'true'` non ?

Comment (dmo-odoo) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2373]: Bah je trouve ça moins bizarre que `attribute === '1'` personellement ? De toute façon techniquement tu peux juste faire `hasAttribute` en fait, non ? Tu t'en fous de sa valeur. En fait je propose de set l'attribut à `''`, ça va mettre l'attribut mais sans valeur, ainsi après tu dois juste faire `hasAttribute`. Ce sera mieux.

Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2373]: ok je vais tester ça.
Attribut sans valeur.

J'espère que ça va pas être cleané par le browser, mais je pense pas.

Comment (sebgeelen) [reply addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:2373]: Seems perfect in fact : https://html.spec.whatwg.org/multipage/common-microsyntaxes.html#boolean-attributes

**Comment by dmo-odoo** [comment]
@robodoo rebase-ff r+
